internal/controllers/machine/machine_controller_inplace_update.go

```diff
@@ -0,0 +1,266 @@
+		s.updatingMessage = "In-place update not possible: InfraMachine not found"
+		return ctrl.Result{}, errors.New("in-place update failed: InfraMachine not found")
+	}
+
+	infraReady := r.isInfraMachineReadyForUpdate(s)
+	bootstrapReady := r.isBootstrapConfigReadyForUpdate(s)
+
+	if !infraReady || !bootstrapReady {
+		log.Info("Waiting for InfraMachine and BootstrapConfig to be marked for in-place update")
+		return ctrl.Result{}, nil
+	}
+
+	log.Info("UpdateMachine hook is pending, calling runtime hook")
+	result, message, err := r.callUpdateMachineHook(ctx, s)
+	if err != nil {
+		s.updatingReason = clusterv1.MachineInPlaceUpdateFailedReason
+		s.updatingMessage = "UpdateMachine hook failed: please check controller logs for errors"
+		return ctrl.Result{}, errors.Wrap(err, "in-place update failed")
+	}
+
+	if result.RequeueAfter > 0 {
+		s.updatingReason = clusterv1.MachineInPlaceUpdatingReason
+		if message != "" {
+			s.updatingMessage = fmt.Sprintf("In-place update in progress: %s", message)
+		} else {
+			s.updatingMessage = "In-place update in progress"
+		}
+		return result, nil
+	}
+
+	log.Info("In-place update completed successfully")
+	if err := r.completeInPlaceUpdate(ctx, s); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to complete in-place update")
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// isInfraMachineReadyForUpdate checks if the InfraMachine has the in-place update annotation.
+func (r *Reconciler) isInfraMachineReadyForUpdate(s *scope) bool {
+	_, hasAnnotation := s.infraMachine.GetAnnotations()[clusterv1.UpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// isBootstrapConfigReadyForUpdate checks if the BootstrapConfig has the in-place update annotation.
+func (r *Reconciler) isBootstrapConfigReadyForUpdate(s *scope) bool {
+	if s.bootstrapConfig == nil {
+		return true
+	}
+	_, hasAnnotation := s.bootstrapConfig.GetAnnotations()[clusterv1.UpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// callUpdateMachineHook calls the UpdateMachine runtime hook for the machine.
+func (r *Reconciler) callUpdateMachineHook(ctx context.Context, s *scope) (ctrl.Result, string, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Validate that exactly one extension is registered for the UpdateMachine hook.
+	// For the current iteration, we only support a single extension to ensure safe behavior.
+	// Support for multiple extensions will be introduced in a future iteration.
+	extensions, err := r.RuntimeClient.GetAllExtensions(ctx, runtimehooksv1.UpdateMachine, s.machine)
+	if err != nil {
+		return ctrl.Result{}, "", err
+	}
+
+	if len(extensions) == 0 {
+		return ctrl.Result{}, "", errors.New("no extensions registered for UpdateMachine hook")
+	}
+
+	if len(extensions) > 1 {
+		return ctrl.Result{}, "", errors.Errorf("multiple extensions registered for UpdateMachine hook: only one extension is supported in the current iteration, found %d extensions: %v", len(extensions), extensions)
```

Comment:
nit, can be fixed also in a follow-up)
 
```suggestion
		return ctrl.Result{}, "", errors.Errorf("multiple extensions registered for UpdateMachine hook: only one extension is supported, found %d extensions: %v", len(extensions), extensions)
```

(not sure if end users care about our iterations; same probably apply to other hooks we introduced in this effort)